### PR TITLE
fix(polars): use `flatten` API for `ArrayFlatten` implementation to avoid large string upcast

### LINKED ISSUE #9995: bug: flatten does not work with polars
### What happened?

I'm taking the example from the [docs](https://ibis-project.org/reference/expression-collections#examples-6) for the `flatten` operation. I'm just adding the line to set the backend to `polars`

```python
import ibis
import ibis.selectors as s
from ibis import _

# Use polars
ibis.set_backend("polars")

ibis.options.interactive = True
schema = {
     "empty": "array<array<int>>",
     "happy": "array<array<string>>",
     "nulls_only": "array<array<struct<a: array<string>>>>",
     "mixed_nulls": "array<array<string>>",
 }
data = {
     "empty": [[], [], []],
     "happy": [[["abc"]], [["bcd"]], [["def"]]],
     "nulls_only": [None, None, None],
     "mixed_nulls": [[], None, [None]],
 }
import pyarrow as pa
t = ibis.memtable(
     pa.Table.from_pydict(
         data,
         schema=ibis.schema(schema).to_pyarrow(),
     )
)

t.happy.flatten()
```

I expected to see:
```
┏━━━━━━━━━━━━━━━━━━━━━┓
┃ ArrayFlatten(happy) ┃
┡━━━━━━━━━━━━━━━━━━━━━┩
│ array<string>       │
├─────────────────────┤
│ ['abc']             │
│ ['bcd']             │
│ ['def']             │
└─────────────────────┘
```

But instead I see the following error:
`ArrowNotImplementedError: Unsupported cast from large_list<item: large_string> to utf8 using function cast_string`

After some investigation, I think that the issue comes from the [polars translator](https://github.com/ibis-project/ibis/blob/54888968db818971c6db7e758f20c91a70f37abb/ibis/backends/polars/compiler.py#L996) for the operation ArrayFlatten:
```python
@translate.register(ops.ArrayFlatten)
def array_flatten(op, **kw):
    return pl.concat_list(translate(op.arg, **kw))
```

Changing the implementation the following seems to solve the issue and produces the expected result
```python
@translate.register(ops.ArrayFlatten)
def array_flatten(op, **kw):
    return translate(op.arg, **kw).flatten()
```

### What version of ibis are you using?

9.3.0

### What backend(s) are you using, if any?

polars 1.5.0

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by cpcloud**
Closes #9995.

### CODE REVIEW

File: ibis/backends/polars/compiler.py [hunk 1]
```diff
@@ -1014,7 +1014,14 @@ def array_collect(op, in_group_by=False, **kw):
...
 @translate.register(ops.ArrayFlatten)
 def array_flatten(op, **kw):
-    return pl.concat_list(translate(op.arg, **kw))
+    result = translate(op.arg, **kw)
+    return (
+        pl.when(result.is_null())
+        .then(None)
+        .when(result.list.len() == 0)
+        .then([])
+        .otherwise(result.flatten())
```
Comment (cpcloud) [inline ibis/backends/polars/compiler.py:1023]: The shenanigans here are because of https://github.com/pola-rs/polars/issues/18522.

File: ibis/backends/tests/test_array.py [hunk 2]
```diff
@@ -1061,9 +1056,9 @@ def test_array_flatten(backend, flatten_data, column, expected):
     data = flatten_data[column]
     t = ibis.memtable({column: data["data"]}, schema={column: data["type"]})
     expr = t[column].flatten()
-    result = backend.connection.execute(expr)
+    result = backend.connection.to_pyarrow(expr)
```
Comment (jcrist) [inline ibis/backends/tests/test_array.py:1059]: Why this change here? Does `con.to_pandas` not work, but `con.to_pyarrow().to_pandas()` does work? If so, we should fix that.

**Review by jcrist — APPROVED** [review]

File: ibis/backends/tests/test_array.py (hunk 2 above)
Comment (cpcloud) [reply ibis/backends/tests/test_array.py:1059]: It works, but `to_pyarrow` reproduces the issue due to polars having different typing behavior for `concat_list` versus `flatten`.

File: ibis/backends/polars/tests/test_client.py [hunk 3]
```diff
@@ -22,3 +25,14 @@ def test_cannot_run_sql_after_drop(con):
     con.drop_table(name)
     with pytest.raises(PolarsSQLInterfaceError):
         con.sql(sql)
+
+
+def test_array_flatten(con):
```
Comment (cpcloud) [inline ibis/backends/polars/tests/test_client.py:30]: Moved this here in order to avoid dealing with Snowflake's PyArrow array output, which differs from other backends because of the untyped nature of its arrays.
